src/librbd/cache/pwl/AbstractWriteLog.cc

```diff
@@ -581,18 +580,20 @@ template <typename I>
 void AbstractWriteLog<I>::update_image_cache_state(Context *on_finish) {
   ldout(m_image_ctx.cct, 10) << dendl;
 
-  ceph_assert(ceph_mutex_is_locked_by_me(m_lock));
-  m_cache_state->allocated_bytes = m_bytes_allocated;
-  m_cache_state->cached_bytes = m_bytes_cached;
-  m_cache_state->dirty_bytes = m_bytes_dirty;
-  m_cache_state->free_bytes = m_bytes_allocated_cap - m_bytes_allocated;
-  m_cache_state->hits_full = m_perfcounter->get(l_librbd_pwl_rd_hit_req);
-  m_cache_state->hits_partial = m_perfcounter->get(l_librbd_pwl_rd_part_hit_req);
-  m_cache_state->misses = m_perfcounter->get(l_librbd_pwl_rd_req) -
-      m_cache_state->hits_full - m_cache_state->hits_partial;
-  m_cache_state->hit_bytes = m_perfcounter->get(l_librbd_pwl_rd_hit_bytes);
-  m_cache_state->miss_bytes = m_perfcounter->get(l_librbd_pwl_rd_bytes) -
-      m_cache_state->hit_bytes;
+  {
+    std::lock_guard locker(m_lock);
+    m_cache_state->allocated_bytes = m_bytes_allocated;
+    m_cache_state->cached_bytes = m_bytes_cached;
+    m_cache_state->dirty_bytes = m_bytes_dirty;
+    m_cache_state->free_bytes = m_bytes_allocated_cap - m_bytes_allocated;
+    m_cache_state->hits_full = m_perfcounter->get(l_librbd_pwl_rd_hit_req);
+    m_cache_state->hits_partial = m_perfcounter->get(l_librbd_pwl_rd_part_hit_req);
+    m_cache_state->misses = m_perfcounter->get(l_librbd_pwl_rd_req) -
+        m_cache_state->hits_full - m_cache_state->hits_partial;
+    m_cache_state->hit_bytes = m_perfcounter->get(l_librbd_pwl_rd_hit_bytes);
+    m_cache_state->miss_bytes = m_perfcounter->get(l_librbd_pwl_rd_bytes) -
+        m_cache_state->hit_bytes;
+  }
   m_cache_state->write_image_cache_state(on_finish);
```

Comment:
I think dropping `m_lock` here may lead to inconsistent cache state blobs (e.g. some counters not summing up correctly or empty/clean booleans being out of sync.  I would prefer if `image_cache_json` string was generated under `m_lock` -- in the same critical section.  Then, `write_image_cache_state` could just take that string knowing that it is consistent even if slightly outdated due to `m_lock` not being held all the way.

Reply:
Otherwise, your approach seems fine to me.  It just needs to be tweaked from "update with m_lock, generate+write without m_lock" to "update+generate with m_lock, write without m_lock".